core/src/main/java/org/mule/work/ProcessingMuleEventWork.java

```diff
@@ -55,7 +59,7 @@ protected void doRun()
     {
         try
         {
-            this.resultEvent = this.messageProcessor.process(event);
+            this.resultEvent = messageProcessorExecutionTemplate.execute(messageProcessor, event);
```

Comment:
Only MuleExceptions are stored for this work instance. For others it is left null.
When getResult checks for exceptions, if it was any other than a MessageException, it would assume it run successfully and return null (both exception and resultEvent are null for this work instance) effectively swallowing the exception.

What the template does, is wrap any exception in a MessagingException (a subclass of MuleException) if necessary, so it is properly handled.